File: mainwindow.cpp
#include "mainwindow.h"
#include "ui_mainwindow.h"
#include "telemetrydata.h"
#include "homepage.h"

// (stubs pages, tu les crées pareil que HomePage)
#include "navigationpage.h"
#include "camerapage.h"
#include "settingspage.h"

#include <QStackedWidget>

MainWindow::MainWindow(TelemetryData* telemetry, QWidget* parent)
    : QMainWindow(parent), ui(new Ui::MainWindow), m_t(telemetry)
{
    ui->setupUi(this);

    // Create pages
    m_home = new HomePage(this);
    m_nav = new NavigationPage(this);
    m_cam = new CameraPage(this);
    m_settings = new SettingsPage(this);

    // Bind telemetry
    m_home->bindTelemetry(m_t);
    m_nav->bindTelemetry(m_t);
    m_cam->bindTelemetry(m_t);
    m_settings->bindTelemetry(m_t);

    // Put into stacked widget
    ui->stackedPages->addWidget(m_home);
    ui->stackedPages->addWidget(m_nav);
    ui->stackedPages->addWidget(m_cam);
    ui->stackedPages->addWidget(m_settings);
    ui->stackedPages->setCurrentWidget(m_home);

    // Bottom nav buttons
    connect(ui->btnHome, &QPushButton::clicked, this, &MainWindow::goHome);
    connect(ui->btnNav, &QPushButton::clicked, this, &MainWindow::goNav);
    connect(ui->btnCam, &QPushButton::clicked, this, &MainWindow::goCam);
    connect(ui->btnSettings, &QPushButton::clicked, this, &MainWindow::goSettings);

    // quick links from home
    connect(m_home, &HomePage::requestNavigation, this, &MainWindow::goNav);
    connect(m_home, &HomePage::requestCamera, this, &MainWindow::goCam);

    // Update topbar + alert
    updateTopBarAndAlert();
    connect(m_t, &TelemetryData::speedKmhChanged, this, &MainWindow::updateTopBarAndAlert);
    connect(m_t, &TelemetryData::batteryPercentChanged, this, &MainWindow::updateTopBarAndAlert);
    connect(m_t, &TelemetryData::gpsOkChanged, this, &MainWindow::updateTopBarAndAlert);
    connect(m_t, &TelemetryData::reverseChanged, this, &MainWindow::updateTopBarAndAlert);
    connect(m_t, &TelemetryData::alertLevelChanged, this, &MainWindow::updateTopBarAndAlert);
    connect(m_t, &TelemetryData::alertTextChanged, this, &MainWindow::updateTopBarAndAlert);
}

MainWindow::~MainWindow(){ delete ui; }

void MainWindow::goHome(){ ui->stackedPages->setCurrentWidget(m_home); }
void MainWindow::goNav(){ ui->stackedPages->setCurrentWidget(m_nav); }
void MainWindow::goCam(){ ui->stackedPages->setCurrentWidget(m_cam); }
void MainWindow::goSettings(){ ui->stackedPages->setCurrentWidget(m_settings); }

void MainWindow::updateTopBarAndAlert(){
    ui->lblSpeed->setText(QString("%1 km/h").arg((int)qRound(m_t->speedKmh())));
    ui->lblBattery->setText(QString("Bat %1%").arg(m_t->batteryPercent()));
    ui->lblGps->setText(m_t->gpsOk() ? "GPS OK" : "GPS LOST");
    ui->lblGear->setText(m_t->reverse() ? "R" : "D");

    if(m_t->alertLevel() == 0){
        ui->alertFrame->setVisible(false);
    } else {
        ui->alertFrame->setVisible(true);
        ui->lblAlertTitle->setText(m_t->alertLevel()==2 ? "CRITIQUE" : "ALERTE");
        ui->lblAlertText->setText(m_t->alertText());
        ui->alertFrame->setStyleSheet(m_t->alertLevel()==2
                                          ? "QFrame{background:#8B1E1E;border-radius:10px;} QLabel{color:white;}"
                                          : "QFrame{background:#6B4B16;border-radius:10px;} QLabel{color:white;}"
                                      );
    }
}

File: main_window.py
from __future__ import annotations

from pathlib import Path

from PyQt6 import uic
from PyQt6.QtWidgets import QMainWindow

from pages import CameraPage, HomePage, NavigationPage, SettingsPage
from telemetry_data import TelemetryData


class MainWindow(QMainWindow):
    def __init__(self, telemetry: TelemetryData, parent=None) -> None:
        super().__init__(parent)
        self._telemetry = telemetry
        uic.loadUi(Path(__file__).resolve().parent / "mainwindow.ui", self)

        self._home = HomePage(self)
        self._nav = NavigationPage(self)
        self._cam = CameraPage(self)
        self._settings = SettingsPage(self)

        self._home.bindTelemetry(telemetry)
        self._nav.bindTelemetry(telemetry)
        self._cam.bindTelemetry(telemetry)
        self._settings.bindTelemetry(telemetry)

        self.stackedPages.addWidget(self._home)
        self.stackedPages.addWidget(self._nav)
        self.stackedPages.addWidget(self._cam)
        self.stackedPages.addWidget(self._settings)
        self.stackedPages.setCurrentWidget(self._home)

        self.btnHome.clicked.connect(self.goHome)
        self.btnNav.clicked.connect(self.goNav)
        self.btnCam.clicked.connect(self.goCam)
        self.btnSettings.clicked.connect(self.goSettings)

        self._home.requestNavigation.connect(self.goNav)
        self._home.requestCamera.connect(self.goCam)

        self.updateTopBarAndAlert()
        telemetry.speedKmhChanged.connect(self.updateTopBarAndAlert)
        telemetry.batteryPercentChanged.connect(self.updateTopBarAndAlert)
        telemetry.gpsOkChanged.connect(self.updateTopBarAndAlert)
        telemetry.reverseChanged.connect(self.updateTopBarAndAlert)
        telemetry.alertLevelChanged.connect(self.updateTopBarAndAlert)
        telemetry.alertTextChanged.connect(self.updateTopBarAndAlert)

    def goHome(self) -> None:
        self.stackedPages.setCurrentWidget(self._home)

    def goNav(self) -> None:
        self.stackedPages.setCurrentWidget(self._nav)

    def goCam(self) -> None:
        self.stackedPages.setCurrentWidget(self._cam)

    def goSettings(self) -> None:
        self.stackedPages.setCurrentWidget(self._settings)

    def updateTopBarAndAlert(self) -> None:
        self.lblSpeed.setText(f"{round(self._telemetry.speedKmh)} km/h")
        self.lblBattery.setText(f"Bat {self._telemetry.batteryPercent}%")
        self.lblGps.setText("GPS OK" if self._telemetry.gpsOk else "GPS LOST")
        self.lblGear.setText("R" if self._telemetry.reverse else "D")

        if self._telemetry.alertLevel == 0:
            self.alertFrame.setVisible(False)
            return

        self.alertFrame.setVisible(True)
        critical = self._telemetry.alertLevel == 2
        self.lblAlertTitle.setText("CRITIQUE" if critical else "ALERTE")
        self.lblAlertText.setText(self._telemetry.alertText)
        if critical:
            self.alertFrame.setStyleSheet("QFrame{background:#8B1E1E;border-radius:10px;} QLabel{color:white;}")
        else:
            self.alertFrame.setStyleSheet("QFrame{background:#6B4B16;border-radius:10px;} QLabel{color:white;}")
